capabilities/client/pkg/discovery/generate_test.go

```diff
@@ -0,0 +1,143 @@
+						{
+							GroupVersionResources: []runv1alpha1.QueryGVR{
+								{
+									Group:    testapigroup.SchemeGroupVersion.Group,
+									Resource: "carps",
+									Versions: []string{testapigroup.SchemeGroupVersion.Version},
+								},
+							},
+							Objects: []runv1alpha1.QueryObject{
+								{
+									ObjectReference:    koi,
+									WithAnnotations:    annotations,
+									WithoutAnnotations: map[string]string{},
+								},
+							},
+							PartialSchemas: []runv1alpha1.QueryPartialSchema{
+								{
+									PartialSchema: "partial schema",
+								},
+							},
+						},
+					},
+				},
+			},
+			err: nil,
+		},
+		{
+			description:  "unknown query type",
+			queryTargets: []QueryTarget{unknownQueryType("shark")},
+			err:          fmt.Errorf("unknown QueryTarget type: %T", unknownQueryType("shark")),
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.description, func(t *testing.T) {
+			got, err := QueryTargetsToCapabilityResource(tc.queryTargets)
+
+			if tc.err == nil {
+				// Check QueryGVR fields.
+				if len(got.Spec.Queries[0].GroupVersionResources[0].Name) == 0 {
+					t.Errorf("QueryGVR name: got: %s, but want: a non-empty string", got.Spec.Queries[0].GroupVersionResources[0].Name)
+				}
+				if got.Spec.Queries[0].GroupVersionResources[0].Group != tc.want.Spec.Queries[0].GroupVersionResources[0].Group {
+					t.Errorf("QueryGVR resource group: got: %s, but want: %s", got.Spec.Queries[0].GroupVersionResources[0].Group, tc.want.Spec.Queries[0].GroupVersionResources[0].Group)
+				}
+				if !reflect.DeepEqual(got.Spec.Queries[0].GroupVersionResources[0].Versions, tc.want.Spec.Queries[0].GroupVersionResources[0].Versions) {
+					t.Errorf("QueryGVR versions: got: %+v, but want: %+v", got.Spec.Queries[0].GroupVersionResources[0].Versions, tc.want.Spec.Queries[0].GroupVersionResources[0].Versions)
+				}
+
+				// Check QueryObject fields.
+				if len(got.Spec.Queries[0].Objects[0].Name) == 0 {
+					t.Errorf("QueryObject name: got: %s, but want: a non-empty string", got.Spec.Queries[0].Objects[0].Name)
+				}
+				if got.Spec.Queries[0].Objects[0].ObjectReference != tc.want.Spec.Queries[0].Objects[0].ObjectReference {
+					t.Errorf("QueryObject object reference: got: %+v, but want: %+v", got.Spec.Queries[0].Objects[0].ObjectReference, tc.want.Spec.Queries[0].Objects[0].ObjectReference)
+				}
+				if !reflect.DeepEqual(got.Spec.Queries[0].Objects[0].WithAnnotations, tc.want.Spec.Queries[0].Objects[0].WithAnnotations) {
+					t.Errorf("QueryGVR versions: got: %+v, but want: %+v", got.Spec.Queries[0].Objects[0].WithAnnotations, tc.want.Spec.Queries[0].Objects[0].WithAnnotations)
+				}
+				if !reflect.DeepEqual(got.Spec.Queries[0].Objects[0].WithoutAnnotations, tc.want.Spec.Queries[0].Objects[0].WithoutAnnotations) {
+					t.Errorf("QueryGVR versions: got: %+v, but want: %+v", got.Spec.Queries[0].Objects[0].WithoutAnnotations, tc.want.Spec.Queries[0].Objects[0].WithoutAnnotations)
+				}
```

Comment:
optional: consider using upstream [diff utilities](https://github.com/kubernetes/utils/blob/cfd413dd9e8559bc69d4abd23c12d666a59971c4/diff/diff.go#L81-L82) to print a human readable diff here. It's hard to know what changed when comparing objects/maps/slices.